### PR TITLE
fix: Hiding widgets / links from unauthorized users on welcome screen.

### LINKED ISSUE #1074: fix: Action link Widgets displayed on main page to unauthorised users.
## This is:

- Bug

## Specifications

  - Browser:

## Expected Behavior

If an account has no permissions to either of these widgets / links I would think they should not be displayed.

Screenshot 2020-06-26 at 12 58 25

## Actual Behavior

They are displayed from an account with no permissions, and when accessing them they report the following error:

Screenshot 2020-06-26 at 13 25 14

## Steps to Reproduce the Problem

1. Create a new user via Security -> Users module, don't assign any roles to the new user.
2. Log in with the new user credentials
2. Open the main navigation and take note of the three widgets.
4. Attempt to navigate to each link.

## Detailed Description

Each panel be hidden if not permitted.

## Possible Solution

Follow similar approaches from
#957 

Hiding using ```SecureView``` and applying the correct scopes for each, where the widgets might be displayed. For example, this ```welcomeScreenWidget``` file from the ```app-form-builder``` folder:

Screenshot 2020-06-26 at 13 40 24

Would you apply the SecureView jsx tag to the WelcomeScreenWidget file?

Maybe this best for a separate issue, but if one or two are only permitted and as such disrupt the design of the site, we could also alter the design with such scenarios. What grid system does the application use? (i.e., CSS Grid ? UI framework grids? CSS / SASS stylings? )

### PR DESCRIPTION
**Pull request by TommyJackson85**
## Related Issue

Closes #1074

## Your solution

Attempting to hide ```welcomeScreenWidget```s from unauthorised users, as they should not have access to if they're not authorised to use them.\
Using ```app-form-builder``` as a first attempt.

My initial attempt was to do it from each welcomeScreenWidget but I only could manage to hide the link button. Here is what it looked like:

Screenshot 2020-06-28 at 13 19 07

So I then applied SecureView to each mapped out Widget plugin from ```app-plugin-admin-welcome-screen/src/components/Welcome.tsx``` file, while passing in a scopes key from the app-form-builder welcomeScreenWidget plugin.
From both unauthorized and authorized accounts, here are the results as it appears to work:

Screenshot 2020-06-28 at 13 49 00

Screenshot 2020-06-28 at 13 48 00

@EmilK15 what do you think? Should I keep the Secure View from the Welcome.tsx file instead and just pass in Secure View scopes from welcomeScreenWidget? I am still yet to do this to other widgets. Just doing one as a first attempt.

## How Has This Been Tested?
I am testing visually and manually through both authorized and unauthorized accounts

### CODE REVIEW

**Comment by TommyJackson85** [comment]
@EmilK15  I updated my initial post to try something different. :)

File: packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx [hunk 1]
```diff
@@ -19,15 +22,17 @@ const buttonStyle = css({
 const plugin: AdminWelcomeScreenWidgetPlugin = {
     type: "admin-welcome-screen-widget",
     name: "admin-welcome-screen-widget-form-builder",
+    scopes: ROLE_FORMS_EDITOR,
```
Comment (Pavel910) [inline packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx:25]: I think this is an acceptable approach 👍

File: packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx [hunk 2]
```diff
@@ -19,15 +22,17 @@ const buttonStyle = css({
...
     widget: {
         cta: (
-            <Link to="/forms" className={linkStyle}>
-                <ButtonSecondary className={buttonStyle}>Create a new Form</ButtonSecondary>
-            </Link>
+            <SecureView scopes={ROLE_FORMS_EDITOR}>
```
Comment (Pavel910) [inline packages/app-form-builder/src/admin/plugins/welcomeScreenWidget.tsx:28]: You can safely remove this `SecureView`.

File: packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx [hunk 3]
```diff
@@ -124,23 +125,25 @@ const Welcome = () => {
...
-                                                    <p className={widgetDescriptionStyle}>
-                                                        {pl.widget.description}
-                                                    </p>
-                                                </Typography>
-                                                <div className={widgetButtonStyle}>
-                                                    {pl.widget.cta}
-                                                </div>
-                                            </Elevation>
-                                        </Widget>
+                                        <SecureView scopes={pl.scopes}>
```
Comment (Pavel910) [inline packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx:128]: This looks good to me. Let's go with this approach.

**Comment by TommyJackson85** [comment]
Yeah for sure, I was considering considering converting it to a function but it could of been
set as an Object for a reason. each object is mapped out as jsx using map from Welcome.tsx.

I agree with you on removing SecureView on each Widget too. I had no plans to keep applying SecureView to the buttons just the widgets. Just left it there to show what I tried.
Better to just pass each scopes from them instead. I will work on it later.

**Comment by TommyJackson85** [comment]
@Pavel910  @EmilK15  I removed the unnecessary SecureViews from the widget and added the remaining scopes key values to the other displayed widgets. Now on unauthorised user accounts it looks like this:

Screenshot 2020-06-28 at 21 02 59

I think for user accounts who have no permissions set this above message should be removed or edited to something more welcoming and applicable because right now they don't have any actions to perform so it's misleading. 
Maybe it should be for a new issue though.

On a side note, do you where else in the project to test the code? Like automated tests? If they exist I could see they cover the code I wrote..

**Comment by doitadrian** [comment]
@TommyJackson85, for Cypress tests, please check the https://github.com/webiny/webiny-js/blob/master/docs/CONTRIBUTING.md#cypress-tests. 

Let us know if you'll have additional questions.

**Comment by Pavel910** [comment]
@TommyJackson85 regarding an empty Welcome screen - let's do it this way: when you get all plugins, you have an array of scopes required by each particular widget. If the user doesn't have access to any of the widgets, we can print a message to contact the administrator. I think this will be enough for now. 

You can check plugin scopes programmatically, by using `import { hasScopes } from "@webiny/app-security";` and render the message accordingly.

I really like how you try to handle all the different edge cases 🚀 🍻 👏

**Comment by TommyJackson85** [comment]
> @TommyJackson85 regarding an empty Welcome screen - let's do it this way: when you get all plugins, you have an array of scopes required by each particular widget. If the user doesn't have access to any of the widgets, we can print a message to contact the administrator. I think this will be enough for now.
> 
> You can check plugin scopes programmatically, by using `import { hasScopes } from "@webiny/app-security";` and render the message accordingly.
> 
> I really like how you try to handle all the different edge cases 🚀 🍻 👏

@Pavel910 I think i managed to use ```hasScopes``` correctly. :) Can you double check my recent commits? It displays a different messaged for users with no authorization. I manually tested it and it seems to work.

@doitadrian thanks! :) Will check it over the week.

**Comment by TommyJackson85** [comment]
@Pavel910 my apologies, I didn't get hasScopes it right yet.

**Comment by Pavel910** [comment]
Hey @TommyJackson85, sorry for the delay. This is what I think will solve the problem:

```tsx
    // Check if user has at least 1 scope
    const canSeeAnything = getPlugins<AdminWelcomeScreenWidgetPlugin>(
        "admin-welcome-screen-widget"
    ).some(pl => hasScopes(pl.scopes, { forceBoolean: true }));
```

We need to see if a user can see at least 1 of the widgets.  Each `Widget` plugin has `scopes` in form of an array. `Array.some` will tell us if at least 1 element passes the test. And `hasScopes` will process an array of scopes defined in the widget plugin, and tell you whether the use has all of the required scopes. It works in the `AND` manner, meaning that the user must have ALL requested scopes for `hasScopes` to return `true`. 

Let me know if this helps!

**Comment by TommyJackson85** [comment]
i thaught it would be fixed as all checks have passed but i am getting this error or running the app locally. The Webiny loading icon appears then this error appears:

Screenshot 2020-07-02 at 13 50 31
Screenshot 2020-07-02 at 13 50 16

File: packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx [hunk 4]
```diff
@@ -112,10 +122,18 @@ const Welcome = () => {
                     <SimpleFormContent>
                         <ContentTheme>
                             <Cell span={12}>
-                                <Typography use={"headline6"}>
-                                    <p className={pGetStartedStyle}>
-                                        To get started - pick one of the actions below:
-                                    </p>
+                                <Typography use={"headline6"}>      
+                                    <h6>{getPlugins<AdminWelcomeScreenWidgetPlugin>("admin-welcome-screen-widget")}</h6>
```
Comment (Pavel910) [inline packages/app-plugin-admin-welcome-screen/src/components/Welcome.tsx:126]: @TommyJackson85 this line here is causing your React error. You're trying to render an array of objects. Not sure what you wanted to render here, but this is the place to fix.

**Comment by TommyJackson85** [comment]
Oh I see. I was only doing that to test to see what it prints 😬was gonna delete later.

**Comment by TommyJackson85** [comment]
How do I grant permission for a user to use each widget? ( i.e. from an administrative account I want to grant an unauthorized account permissions for each widget access). This is just so I can do further manual testing on an account to get authorization for one or two widgets.

So far from and administrator account, I can only give a user these permissions:
Screenshot 2020-07-02 at 21 19 29

Anyway, I will give the Cyprus tests a shot, unless you want to merge these changes as they are.

**Comment by Pavel910** [comment]
@TommyJackson85 roles contain scopes. So you need to create a Role (or several of them) with a specific set of scopes. This way we allow administrators to structure their permissions however they like, and we only care the scopes at the code level.

**Comment by TommyJackson85** [comment]
Ok I’ll try it later, I need to delete a comment I left too.

**Comment by TommyJackson85** [comment]
I think I managed to grant permissions for page building, but they only seem to appear in the side nav bar :( 
Screenshot 2020-07-04 at 13 01 43

**Comment by Pavel910** [comment]
@TommyJackson85 that's because the scope in widget plugins are not really related to what you have assigned to the user. In the Page Builder widget you have `scopes: ["pb:settings"]`, should be `pb:page:crud`, and so on for all the plugins.

**Comment by TommyJackson85** [comment]
> @TommyJackson85 that's because the scope in widget plugins are not really related to what you have assigned to the user. In the Page Builder widget you have `scopes: ["pb:settings"]`, should be `pb:page:crud`, and so on for all the plugins.

That is fair enough. So 'pb:settings' has nothing to do with the widget plugins? Should I have given different scopes for the widgets instead of 'settings' scopes? Whould I have given a list of all 'pb scopes maybe'?

**Comment by TommyJackson85** [comment]
@Pavel910 would I be better to apply all potential scope types to the the scopes array including both ```pb:settings``` and ```pb:page:crud``` for example?

**Comment by Pavel910** [comment]
@TommyJackson85 I think `pb:page:crud` makes the most sense, because the widget is linking to Pages List, which is a module for creating pages. `pb:settings` is not relevant here. Also since `hasScopes` works in an `AND` fashion, you need to be careful not to include scopes that may not be present in user's token.

**Comment by TommyJackson85** [comment]
Ok so I will just use  page:crud   scopes for now so. thanks for the advice!

**Comment by TommyJackson85** [comment]
**These are the scopes I found, as you see there is nothing for CMS**

Screenshot 2020-07-04 at 19 54 52
Screenshot 2020-07-04 at 19 54 40
Screenshot 2020-07-04 at 19 53 57
Screenshot 2020-07-04 at 19 52 33**
Screenshot 2020-07-04 at 19 50 35
Screenshot 2020-07-04 at 19 50 20

**Comment by Pavel910** [comment]
@TommyJackson85 CMS scopes are in progress in another PR: https://github.com/webiny/webiny-js/pull/1088.
You can safely ignore that one for now, simply set the scopes to `[]` to let that widget always be visible, and we'll come back to it later.

**Comment by TommyJackson85** [comment]
> @TommyJackson85 CMS scopes are in progress in another PR: #1088.
> You can safely ignore that one for now, simply set the scopes to `[]` to let that widget always be visible, and we'll come back to it later.

Ok, no problem!

**Comment by TommyJackson85** [comment]
```pb:page:crud``` and ```forms:form:crud``` are now the scopes being used for their respective plugins and they appear to work as I've tested them manually. CMS Widgets remains accessable for all as its scopes are empty.
Are you happy to merge this PR?

Just for further context, from a previous issue I fixed, I added "settings" scopes for page builder, form builder and headless cms under settings in the side nav. It might be worth retesting them. Should the crud scopes be applied there too?

**Comment by Pavel910** [comment]
@TommyJackson85 Great! I think this is good for merging, and no you don’t need to add crud scopes to settings navigation items. Settings scopes are for settings, crud scopes are for crud modules. So it’s all good now. We’ll do further refinement later on, when the need arises.

I’ll merge this for the next release, feel free to change the draft status of PR and please update the PR body to remove the temporary text.

Thanks and good work! :)

**Comment by TommyJackson85** [comment]
How do I change the draft status? Will be good to learn this.

**Pull request merged**